### PR TITLE
WPF - Implement Touch and Stylus Input

### LINKED ISSUE #228: Feature Request - Add WPF TouchScroll/Multitouch Support
I guess this is more of a feature request than an issue but it would be nice to have multitouch support:  

-Scrolling with one finger
-Zooming with pinch zoom gesture

### PR DESCRIPTION
**Pull request by GSonofNun**
Resolves #228 

Feedback welcome. In general seems to work the same way chrome does with touch and stylus input.

### CODE REVIEW

**Comment by amaitland** [comment]
Thanks :+1:

What versions of windows have you tested on? What touch devices?

> In general seems to work the same way chrome does with touch and stylus input.

How many different machines have you tested on? What testing did you perform?

File: CefSharp.Wpf/Internals/WpfExtensions.cs [hunk 1]
```diff
@@ -72,6 +72,90 @@ public static CefEventFlags GetModifiers(this MouseEventArgs e)
...
+
+            return modifiers;
+        }
+
+        /// <summary>
+        /// Gets the modifiers.
+        /// </summary>
+        /// <param name="e">The <see cref="StylusEventArgs"/> instance containing the event data.</param>
+        /// <returns>CefEventFlags.</returns>
+        public static CefEventFlags GetModifiers(this StylusEventArgs e)
```
Comment (amaitland) [inline CefSharp.Wpf/Internals/WpfExtensions.cs]: Does this actually use `StylusEventArgs`?

Cursors glance it looks like this code is now duplicated three times, maybe I'm missing something subtle?

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 2]
```diff
@@ -2191,7 +2191,7 @@ protected override void OnMouseLeave(MouseEventArgs e)
         /// <param name="e">The <see cref="MouseButtonEventArgs"/> instance containing the event data.</param>
         private void OnMouseButton(MouseButtonEventArgs e)
         {
-            if (!e.Handled && browser != null)
+            if (!e.Handled && e.StylusDevice == null && browser != null)
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: What is the reason for the additional check? Please add a comment explaining.

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 3]
```diff
@@ -2220,6 +2220,150 @@ private void OnMouseButton(MouseButtonEventArgs e)
...
+        /// </summary>
+        /// <param name="e">The <see cref="TouchEventArgs"/> instance containing the event data.</param>
+        private void OnTouch(TouchEventArgs e)
+        {
+            var modifiers = e.GetModifiers();
+            var touchPoint = e.GetTouchPoint(this);
+            var touchEventType = TouchEventType.Cancelled;
+            switch (touchPoint.Action)
+            {
+                case TouchAction.Down:
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Braces for case blocks, even single line ones please.

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 4]
```diff
@@ -2220,6 +2220,150 @@ private void OnMouseButton(MouseButtonEventArgs e)
...
+                    touchEventType = TouchEventType.Cancelled;
+                    break;
+            }
+
+            var touchEvent = new TouchEvent()
+            {
+                Id = e.TouchDevice.Id,
+                X = (float)touchPoint.Position.X,
+                Y = (float)touchPoint.Position.Y,
+                RadiusX = 0,
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Shouldn't this default to `0` already? (And the others like it).

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 5]
```diff
@@ -2220,6 +2220,150 @@ private void OnMouseButton(MouseButtonEventArgs e)
...
+            base.OnTouchUp(e);
+        }
+
+        /// <summary>
+        /// Handles a <see cref="E:Touch" /> event.
+        /// </summary>
+        /// <param name="e">The <see cref="TouchEventArgs"/> instance containing the event data.</param>
+        private void OnTouch(TouchEventArgs e)
+        {
+            var modifiers = e.GetModifiers();
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Should be a similar check to the Mouse handlers, `if (!e.Handled && browser != null)`

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 6]
```diff
@@ -2220,6 +2220,150 @@ private void OnMouseButton(MouseButtonEventArgs e)
...
+        }
+
+        /// <summary>
+        /// Handles a <see cref="E:Stylus" /> event.
+        /// </summary>
+        /// <param name="e">The <see cref="StylusEventArgs"/> instance containing the event data.</param>
+        /// <param name="touchEventType">The <see cref="TouchEventType"/> event type</param>
+        private void OnStylus(StylusEventArgs e, TouchEventType touchEventType)
+        {
+            var stylusPoints = e.GetStylusPoints(this);
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Again similar `if (!e.Handled && browser != null)` check

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 7]
```diff
@@ -2220,6 +2220,150 @@ private void OnMouseButton(MouseButtonEventArgs e)
...
+            if (e.StylusDevice != null && e.StylusDevice.TabletDevice.Type == TabletDeviceType.Stylus && stylusPoints.Count > 0)
+            {
+                var modifiers = e.GetModifiers();
+                var pressure = stylusPoints[0].PressureFactor;
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Looks like you assign a `pressure` variable, doesn't appear to be used though.

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 8]
```diff
@@ -2220,6 +2220,150 @@ private void OnMouseButton(MouseButtonEventArgs e)
...
         }
 
+        /// <summary>
+        /// Provides class handling for the <see cref="E:System.Windows.TouchDown" /> routed event that occurs when a touch presses inside this element.
+        /// </summary>
+        /// <param name="e">The <see cref="T:System.Windows.Input.TouchEventArgs" /> that contains the event data.
+        protected override void OnTouchDown(TouchEventArgs e)
+        {
+            Focus();
+            CaptureTouch(e.TouchDevice);
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Please add a code comment explaining the reason for the capture, what use case is it required for.

**Comment by amaitland** [comment]
Your probably aware that the `WPF` touch implementation is known to have performance issues  https://social.msdn.microsoft.com/Forums/vstudio/en-US/bce8c2f4-a7aa-40c3-a28e-8f03f78dfd4b/touch-event-delaylag?forum=wpf

One possible workaround is to use https://github.com/jaytwo/WmTouchDevice we should probably consider adding that to the example `WPF` app.

**Comment by mortalV** [comment]
Just my 5 cents about WmTouchDevice.
We used this solution before, but we had to refuse because some of the events sometimes are lost (for example, TouchUp).
We did not use this solution for CefSharp, but even ordinary controls were strange in terms of generating events.

**Comment by amaitland** [comment]
> Just my 5 cents about WmTouchDevice.

@mortalV How long ago did you perform your testing? I've seen many reports of `WmTouchDevice` resolving touch related issues.

Enabling it in the example would have no bearing on your application, you can choose whatever solution you like.

**Comment by mortalV** [comment]
We have testing just today but maybe you are right and we have old version of WmTouchDevice.
I'll look reports and what was changed in library. Thanks for info.

**Comment by GSonofNun** [comment]
> What versions of windows have you tested on? [...] How many different machines have you tested on?

I have currently only tested it on a Surface Book 2 running Windows 10 Pro Version 1903 Build 18875.1000. I plan to also test it on a Surface Go running Windows 10 Home Version 1803 Build 17134.706. Those are currently the only touch/pen-enabled devices I can test with.

> What touch devices?

I'm testing touch, multi-touch, and pen input.

>  What testing did you perform?

- Touch Scrolling
- Touch "right-click"
- Pinch-to-zoom with multi-touch
- Scrolling with stylus
- Stylus pressure (tested on pressurejs.com)
- Scrolling with Stylus eraser (I couldn't find a website that specifically used an eraser though)

**Comment by GSonofNun** [comment]
> Your probably aware that the WPF touch implementation is known to have performance issues

I was not aware, but it did feel a little laggy when I was testing. I wasn't sure what exactly was causing it, and even mouse-scrolling felt sluggish, so I didn't look much into it.

> One possible workaround is to use https://github.com/jaytwo/WmTouchDevice we should probably consider adding that to the example WPF app.

I'll look into it.

File: CefSharp.Wpf/Internals/WpfExtensions.cs (hunk 1 above)
Comment (GSonofNun) [reply CefSharp.Wpf/Internals/WpfExtensions.cs]: I was trying to mimic the other methods in the class, as I didn't know the expected way of doing similar things. I have updated the code to remove the duplicated code and unused parameters.

File: CefSharp.Wpf/ChromiumWebBrowser.cs (hunk 2 above)
Comment (GSonofNun) [reply CefSharp.Wpf/ChromiumWebBrowser.cs]: Added comment in code. (Hopefully that's where you meant)

File: CefSharp.Wpf/ChromiumWebBrowser.cs (hunk 4 above)
Comment (GSonofNun) [reply CefSharp.Wpf/ChromiumWebBrowser.cs]: Removed unnecessary initializers.

File: CefSharp.Wpf/ChromiumWebBrowser.cs (hunk 5 above)
Comment (GSonofNun) [reply CefSharp.Wpf/ChromiumWebBrowser.cs]: Added if check

File: CefSharp.Wpf/ChromiumWebBrowser.cs (hunk 6 above)
Comment (GSonofNun) [reply CefSharp.Wpf/ChromiumWebBrowser.cs]: Added additional if checks

File: CefSharp.Wpf/ChromiumWebBrowser.cs (hunk 7 above)
Comment (GSonofNun) [reply CefSharp.Wpf/ChromiumWebBrowser.cs]: My mistake. I forgot to use the variable. Fixed.

File: CefSharp.Wpf/ChromiumWebBrowser.cs (hunk 8 above)
Comment (GSonofNun) [reply CefSharp.Wpf/ChromiumWebBrowser.cs]: Added comments explaining my reasoning for the Captures.

**Comment by GSonofNun** [comment]
@amaitland I don't know the protocol for resolving change requests. Do I mark comments/change-requests as resolved when I think I have resolved them? Or do I leave them unresolved for the change-requestor to resolve?

**Comment by amaitland** [comment]
> Do I mark comments/change-requests as resolved when I think I have resolved them?

@GSonofNun You can mark them as resolved.

**Comment by amaitland** [comment]
My overall concern with this is testing, code it's self looks clean and concise 👍 

To get things moving forward and making it easy for others to test I'll create an `CefSharp.Wpf.Experimental.ChromiumWebBrowserWithTouchSupport` class and include your code changes in the upcoming `73` release (it's way too late in the development cycle to be including the changes by default). This should allow others to easily test your changes in a variety of hardware, we still need people to test `Windows 7` and `Windows 8.1`.

**Comment by amaitland** [comment]
Very quickly added `ChromiumWebBrowserWithTouchSupport` in https://github.com/cefsharp/CefSharp/commit/3d14f198674dba6516276897d384c1cc0c996f92

I haven't had time to do much more than compile it at this point in time, feedback welcome, hopefully I haven't missed anything.

**Comment by amaitland** [comment]
> Very quickly added `ChromiumWebBrowserWithTouchSupport` in [3d14f19](https://github.com/cefsharp/CefSharp/commit/3d14f198674dba6516276897d384c1cc0c996f92)

Changes were included as part of `73.1.130`, anyone interested can use `ChromiumWebBrowserWithTouchSupport` to test this out.

Closing this now, any further changes please open a follow up `PR` targeted against `master`. Thanks for the valuable contribution 👍

**Pull request closed without merging**